sstables/storage.cc

```diff
@@ -596,17 +614,35 @@ void s3_storage::open(sstable& sst) {
 }
 
 future<file> s3_storage::open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) {
-    co_return _client->make_readable_file(make_s3_object_name(sst, type), _as);
+    return maybe_wrap_file(sst, type, flags, _client->make_readable_file(make_s3_object_name(sst, type), _as));
+}
+
+static future<data_sink> maybe_wrap_sink(const sstable& sst, component_type type, data_sink sink) {
+    if (type != component_type::TOC && type != component_type::TemporaryTOC) {
+        for (auto* ext : sst.manager().config().extensions().sstable_file_io_extensions()) {
+            std::exception_ptr p;
+            try {
+                sink = co_await ext->wrap_sink(sst, type, std::move(sink));
+            } catch (...) {
+                p = std::current_exception();
+            }
+            if (p) {
+                co_await sink.close();
+                std::rethrow_exception(std::move(p));
+            }
+        }
+    }
+    co_return sink;
 }
 
 future<data_sink> s3_storage::make_data_or_index_sink(sstable& sst, component_type type) {
     SCYLLA_ASSERT(type == component_type::Data || type == component_type::Index);
     // FIXME: if we have file size upper bound upfront, it's better to use make_upload_sink() instead
-    co_return _client->make_upload_jumbo_sink(make_s3_object_name(sst, type), std::nullopt, _as);
+    return maybe_wrap_sink(sst, type, _client->make_upload_jumbo_sink(make_s3_object_name(sst, type), std::nullopt, _as));
 }
 
 future<data_sink> s3_storage::make_component_sink(sstable& sst, component_type type, open_flags oflags, file_output_stream_options options) {
-    co_return _client->make_upload_sink(make_s3_object_name(sst, type), _as);
+    return maybe_wrap_sink(sst, type, _client->make_upload_sink(make_s3_object_name(sst, type), _as));
```

Comment:
So, in the end of the day, we have:

* For file storage
  * `open_component` returns posix file that's wrapped in all extensions out there
  * `make_component_sink`/`make_data_or_index_sink` both return file_data_sink that references file wrapped by `open_component`
* For S3 storage
  * `open_component` returns s3 readable file that's wrapped in all extensions out there
  * `make_component_sink`/`make_data_or_index_sink` both return file_data_sink that references sink_file_impl that references s3 upload sink(s)

Is that summary correct?

Reply:
> No, the last patches makes sure the data sinks are only wrapped in a single level of sink, no file objects in-between. 

Only for S3, or file storage as well?

Reply:
For file storage, we always use the product of open_component, thus extensions are already applied.